enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,339 @@
+case where more than an ipv4 and ipv6 address would be needed.
+
+### openshift/api
+
+We will need to make changes to the
+[baremetal platformStatus](https://github.com/openshift/api/blob/354aa98a475c1fcd60b41aaee735da95d7318b42/config/v1/0000_10_config-operator_01_infrastructure.crd.yaml#L311)
+to add new fields for the new VIPs. Unlike the install-config layout, I think
+it is clearly better to add discrete new fields for this feature. That way,
+backend code in machine-config-operator and baremetal-runtimecfg will not
+need to detect whether the values in the old fields are from a new (where
+the field will have to be v4) or old (where the existing field can be either
+v4 or v6) configuration.
+
+Ideally we would transition existing values to the new fields on upgrade.
+I'm unsure whether a mechanism exists for us to do that, however. If not, then
+some logic (probably in MCO) will be needed to map values from the old fields
+to the appropriate ip version.
+
+The new structure of platformStatus would look like this:
+
+```
+baremetal:
+  description: BareMetal contains settings specific to the BareMetal platform.
+  type: object
+  properties:
+    # Deprecated
+    apiServerInternalIP:
+      description: apiServerInternalIP is an IP address...
+      type: string
+    apiServerInternalIPv4:
+      description: apiServerInternalIPv4 is an IP address...
+      type: string
+    apiServerInternalIPv6:
+      description: apiServerInternalIPv6 is an IP address...
+      type: string
+    # Deprecated
+    ingressIP:
+      description: ingressIP is an external IP...
+      type: string
+    ingressIPv4:
+      description: ingressIPv4 is an external IP...
+      type: string
+    ingressIPv6:
+      description: ingressIPv6 is an external IP...
+      type: string
+    nodeDNSIP:
+    ...
+```
+
+### machine-config-operator
+
+We will need to wire in the new VIPs to the manifests and configuration
+templates. There should be conditionals around the v4 and v6 sections
+so only the appropriate ones are enabled based on the provided VIPs.
+
+I believe haproxy already listens on both ipv4 and ipv6 so there should be no
+changes needed in that configuration. We'll need to verify that as part of
+this implementation though. Note that as of this writing, the apiserver
+does not support dual stack anyway so all traffic through haproxy will end
+up on one or the other ip version.
+
+As noted above, it would be best to migrate the existing VIP fields to the new
+v4 and v6 fields to simplify the logic in consumers. We may be able to do that
+in the
+[merge code run on upgrade](https://github.com/openshift/machine-config-operator/blob/080919dc992d600706b679eb118393ee293496f7/lib/resourcemerge/machineconfig.go#L68).
+I'm unsure whether that can modify the actual infrastructure object, but
+it can massage the data put into controllerconfig, which is where we actually
+get the values.
```

Comment:
Right yes, the controllerconfig is just an internal implementation detail of the MCO we can change at will.